thriftbp/client_middlewares.go

```diff
@@ -313,3 +318,81 @@ var (
 	_ thrift.ClientMiddleware = SetDeadlineBudget
 	_ thrift.ClientMiddleware = BaseplateErrorWrapper
 )
+
+// PrometheusClientMiddleware returns middleware to track Prometheus metrics
+// specific to the Thrift client.
+//
+// It emits the following prometheus metrics:
+//
+// * thrift_client_active_requests gauge with labels:
+//
+//   - thrift_service: the serviceSlug arg
+//   - thrift_method: the method of the endpoint called
+//   - thrift_slug: the service being contacted
+//
+// * thrift_client_latency_seconds histogram with labels above plus:
+//
+//   - thrift_success: "true" if err == nil, "false" otherwise
+//
+// * thrift_client_requests_total counter with all labels above plus:
+//
+//   - thrift_exception_type: the human-readable exception type, e.g.
+//     baseplate.Error, etc
+//   - thrift_baseplate_status: the numeric status code from a baseplate.Error
+//     as a string if present (e.g. 404), or the empty string
+//   - thrift_baseplate_status_code: the human-readable status code, e.g.
+//     NOT_FOUND, or the empty string
+func PrometheusClientMiddleware(localServiceSlug, remoteServiceSlug string) thrift.ClientMiddleware {
+	return func(next thrift.TClient) thrift.TClient {
+		return thrift.WrappedTClient{
+			Wrapped: func(ctx context.Context, method string, args, result thrift.TStruct) (_ thrift.ResponseMeta, err error) {
+				start := time.Now()
+				activeRequestLabels := prometheus.Labels{
+					localServiceLabel:      localServiceSlug,
+					methodLabel:            method,
+					remoteServiceSlugLabel: remoteServiceSlug,
+				}
+				clientActiveRequests.With(activeRequestLabels).Inc()
+
+				defer func() {
+					var exceptionTypeLabel, baseplateStatusCode, baseplateStatus string
+					success := strconv.FormatBool(err == nil)
+					if err != nil {
+						exceptionTypeLabel = strings.TrimPrefix(fmt.Sprintf("%T", err), "*")
```

Comment:
Hmm, this has the potential to be a huge un-bounded cardinality bomb. I would highly recommend not including something like this.

Reply:
oh yeah. good catch. let me look into a fix

Reply:
So I had a chat with @kylelemons. Apparently the `%T` limits it to just the error type. It still concerns me, but maybe it's misplaced.

Reply:
Kyle pointed this out to me earlier today but this probably isn't actually all that useful, any error that isn't an actual struct will just show up as `errors.errorString`:

https://play.golang.org/p/bEbqYiP4dL1

Reply:
Ah, OK so Thrift is in a bit of a different place.  Thrift errors _do_ have to be specific error types, because that's how Thrift decides what to send over the wire.  So, it actually _is_ meaningful in this case.  For Thrift, this is basically how you ask what type of Exception was thrown.

Reply:
Note this would probably end up having a similar issue since we've been encouraging devs to use the `baseplate.Error` struct rather than defining their own exceptions. On the plus side though, you could see if it _is_ a `baseplate.Error` and extract the error code and maybe turn that into a string and put it in there as `baseplate.Error.${code}`?